rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,476 @@
+- Context is harder to discover as it is outside the props contract.
+- Callers doing custom styling may still take a dependency on the CSS internals of the component if they don't replace a
+  style block completely.
+
+### 🤔 Centralized and singular custom styling method
+
+It would be nice to have a single method on context that could apply for all components. This proves difficult
+with each component having its own props and state types.
+
+Another approach would be to define single context value type in the react-shared-contexts package that had
+a method for each component, but that causes a circular dependency problem for components.
+
+There may some typescript trickery that would allow for a single context value type, but having each component define
+the context type it will inspect for custom styling follows SOLID dependency inversion.
+
+## Option B: Encourage recomposition | make recomposition easier
+
+The hooks composition model was architected to separate concerns of component behavior, style, and rendering.
+Rather than provide an additional mechanism for custom styling, this option would assist partners with leveraging
+existing hooks to recompose components with a different styling mechanism.
+
+Each component would be recomposed.
+
+For example, imagine this Button.tsx is defined in a partner component library to export a custom styled button.
+The component exported is named `Button` so that it can replace Fluent UI's `Button`.
+
+The partner would define the useCustomButtonStyles hooks to style the component however they like.
+This could include exposing a similar centralized custom styling mechanism as in Option A or applying styles
+through a different mechanism than Griffel (such as SASS, emotion, or tailwind).
+
+```tsx
+import * as React from 'react';
+import { renderButton_unstable, useButton_unstable } from '@fluentui/react-components';
+import type { ButtonProps, ForwardRefComponent } from '@fluentui/react-components';
+import { useCustomButtonStyles } from '../styling/button';
+
+export const Button: ForwardRefComponent<ButtonProps> = React.forwardRef((props, ref) => {
+  const state = useButton_unstable(props, ref);
+  useCustomButtonStyles(state);
+  return renderButton_unstable(state);
+}) as ForwardRefComponent<ButtonProps>;
+
+Button.displayName = 'Button';
+```
+
+### 💡 Improvement - making recomposition easier with compose()
+
+Recomposing every component could be made easier with a compose method that hid some of the details of the ForwardRef
+and sequencing the hook.
+
+```tsx
+import * as React from 'react';
+import { compose } from '@fluentui/react-components';
+import type { ButtonProps } from '@fluentui/react-components';
+import { useCustomButtonStyles } from '../styling/button';
+
+// This is a rough idea of what a compose method might look like.
+export const Button = compose<ButtonProps>('Button', useButton_unstable, useCustomButtonStyles, renderButton_unstable);
```

Comment:
If we want to simplify styles and hide details, why we want to have bits not related to styling in API?

```js
export const Button = compose<ButtonProps>(Button, useCustomButtonStyles);
```

Should it be just something like that instead?

Reply:
Ben has been hard at work refining the composition/recomposition helpers.  He can propose and PR that separately.  I think this details the idea enough that we can evaluate option B as the 'don't build a separate escape-hatch for styling if we are all-in on recomposition'.